lib/spack/spack/util/atomic_update.py

```diff
@@ -0,0 +1,79 @@
+# Copyright 2013-2023 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+import ctypes
+import os
+from contextlib import contextmanager
+from typing import Optional
+
+import llnl.util.filesystem as fs
+from llnl.util.symlink import symlink
+
+libc: Optional[ctypes.CDLL] = None
+try:
+    libc = ctypes.CDLL("/lib64/libc.so.6", 0x04)  # 0x04 is RTLD_NOLOAD
+except BaseException:
+    pass
+
+
+def use_renameat2():
+    return hasattr(libc, "renameat2")
+
+
+def atomic_update(oldpath, newpath):
+    """
+    atomically update newpath to contain the information at oldpath
+
+    on linux systems supporting renameat2, the paths are swapped.
+    on other systems, oldpath is not affected but all paths are abstracted
+    by a symlink to allow for atomic updates.
+    """
+    if use_renameat2():
+        return atomic_update_renameat2(oldpath, newpath)
+    else:
+        return atomic_update_symlink(oldpath, newpath)
+
+
+@contextmanager
+def open_safely(path):
+    fd = os.open(path, os.O_CLOEXEC | os.O_PATH)
+    try:
+        yield fd
+    finally:
+        os.close(fd)
+
+
+def atomic_update_renameat2(src, dest):
+    dest_exists = os.path.exists(dest)
+    if not dest_exists:
+        fs.touch(dest)
+    with open_safely(src) as srcfd:
+        with open_safely(dest) as destfd:
+            try:
+                libc.renameat2(
+                    srcfd, src.encode(), destfd, dest.encode(), 2
```

Comment:
srcfd and destfd are ignored if you pass absolute paths, can you drop that `open_safely` business then and replace by abspath?